3.next - Integrate CookieCollection into Response

markstory opened the pull request:
Integrate the new Cookies and CookieCollection into the Response. All of the existing cookie related methods are backwards compatible, and `withCookie()` allows new style cookies as a parameter.

I've had to add more compatibility shims into Cookie and relax the expires parameter typing to allow integers. I found this preferable to coercing the existing cookie data into a DateTime just so it can be converted back into an integer.

Refs #10406

markstory committed:
Tidy up class comment.

Remove useless information.

markstory committed:
Add another compatibility shim for cookies on the response.

Rename the existing compat method as there are two now.

markstory committed:
Update Response to use CookieCollection internally.

This change should be completely backwards compatible and enable the new
cookie collection to be used.

markstory committed:
Fix tests failing from invalid cookie names.

markstory committed:
Allow withCookie() to take cookie objects.

Seems like a reasonable thing to do.

robertpustulka commented on src/Http/Response.php:
```diff
@@ -462,6 +461,7 @@ public function __construct(array $options = [])
             $this->_contentType = $this->resolveType($options['type']);
         }
         $this->_setContentType();
+        $this->_cookies = new CookieCollection();
```
Can we inject a collection from `$options`?

markstory replied:
Not right now. Responses don't take cookies in their constructor at this time.

robertpustulka commented on src/Http/Response.php:
```diff
@@ -1977,30 +1985,44 @@ public function cookie($options = null)
 
         $new = clone $this;
-        $new->_cookies[$name] = $data;
+        $new->_cookies = $new->_cookies->add($cookie);
```
I think a collection should be cloned as well. Adding a cookie to the new request's collection would also mutate current request's collection.

markstory committed:
Allow spaces in cookie names.

We need this for backwards compatibility. Also make the invalid name
tests more thorough.

markstory committed:
Fix incorrect type matching in Response::cookie()

markstory committed:
Fix typehint

markstory committed:
Use DateTime objects internally.

This will allow expiration dates to exist in the far future on 32bit
(windows) systems.

markstory committed:
Fix PHPCS error.

markstory committed:
Return timestamps as strings.

This prevents overflow issues on 32bit windows systems for far futures
expiration dates.

ravage84 commented on src/Http/Cookie/Cookie.php:
```diff
@@ -251,7 +271,7 @@ public function getName()
     protected function validateName($name)
     {
-        if (preg_match("/[=,; \t\r\n\013\014]/", $name)) {
+        if (preg_match("/[=,;\t\r\n\013\014]/", $name)) {
```
@markstory do you know of any documented list of invalid characters? If so, we could link to it with a ``@link``.

markstory replied:
The cookie RFCs have the accepted character lists.

ravage84 replied:
If I interpreted that correctly:

> cookie-name       = token
> token             = <token, defined in [RFC2616], Section 2.2>

https://tools.ietf.org/html/rfc6265#section-4.1.1

So we could add:
``@link https://tools.ietf.org/html/rfc2616#section-2.2 Rules for naming cookies.``

markstory replied:
done.

markstory committed:
Add link to cookie RFC for reference.
